src/EditorFeatures/Core/Implementation/IntelliSense/Completion/FileSystem/FileSystemCompletionHelper.cs

```diff
@@ -6,152 +6,211 @@
+            _baseDirectoryOpt = baseDirectoryOpt;
+            _allowableExtensions = allowableExtensions;
             _folderGlyph = folderGlyph;
             _fileGlyph = fileGlyph;
-            _exclude = exclude;
             _itemRules = itemRules;
 
-            _lazyGetDrives = new Lazy<string[]>(() =>
-                IOUtilities.PerformIO(Directory.GetLogicalDrives, Array.Empty<string>()));
+            _lazyGetDrives = new Lazy<string[]>(GetLogicalDrives);
         }
 
-        public ImmutableArray<CompletionItem> GetItems(string pathSoFar, string documentPath)
+        // virtual for testing
+        protected virtual string[] GetLogicalDrives() => 
+            IOUtilities.PerformIO(CorLightup.Desktop.GetLogicalDrives, Array.Empty<string>());
+
+        // virtual for testing
+        protected virtual bool DirectoryExists(string fullPath)
         {
-            if (_exclude != null && _exclude(pathSoFar))
+            Debug.Assert(PathUtilities.IsAbsolute(fullPath));
+            return Directory.Exists(fullPath);
+        }
+
+        // virtual for testing
+        protected virtual IEnumerable<string> EnumerateDirectories(string fullDirectoryPath)
+        {
+            Debug.Assert(PathUtilities.IsAbsolute(fullDirectoryPath));
+            try
             {
-                return ImmutableArray<CompletionItem>.Empty;
+                return Directory.EnumerateDirectories(fullDirectoryPath);
+            }
+            catch
+            {
+                return Array.Empty<string>();
             }
-
-            return GetFilesAndDirectories(pathSoFar, documentPath);
         }
 
-        private CompletionItem CreateCurrentDirectoryItem()
+        // virtual for testing
+        protected virtual IEnumerable<string> EnumerateFiles(string fullDirectoryPath)
         {
-            return CommonCompletionItem.Create(".", rules: _itemRules);
+            Debug.Assert(PathUtilities.IsAbsolute(fullDirectoryPath));
+            try
+            {
+                return Directory.EnumerateFiles(fullDirectoryPath);
+            }
+            catch
+            {
+                return Array.Empty<string>();
+            }
         }
 
-        private CompletionItem CreateParentDirectoryItem()
+        // virtual for testing
+        protected virtual bool IsVisibleFileSystemEntry(string fullPath)
         {
-            return CommonCompletionItem.Create("..", rules: _itemRules);
+            Debug.Assert(PathUtilities.IsAbsolute(fullPath));
+            try
+            {
+                return (File.GetAttributes(fullPath) & (FileAttributes.Hidden | FileAttributes.System)) == 0;
+            }
+            catch
+            {
+                return false;
+            }
         }
 
-        private CompletionItem CreateNetworkRoot(TextSpan textChangeSpan)
+        private CompletionItem CreateNetworkRoot() =>
+            CommonCompletionItem.Create(
+                "\\\\",
+                glyph: null,
```

Comment:
Yeah, we don't have a good glyph for a network root. It's not really a directory.

---
In reply to: [114853511](https://github.com/dotnet/roslyn/pull/19195#discussion_r114853511) [](ancestors = 114853511)